✨(frontend) add pdf block to the editor

dakshesh14 opened the pull request:
## Purpose

Added an option for users to upload and preview PDF files in the editor. This PR addresses #348. It is a duplicate of [PR-795](https://github.com/suitenumerique/docs/pull/795) because I couldn't reopen the PR as I lost my fork.

[pdf-embed.webm](https://github.com/user-attachments/assets/32aa040b-62ba-4818-92d6-af7c4c99fe5f)

<img width="417" height="900" alt="mobile-view-demo" src="https://github.com/user-attachments/assets/77eea8c4-b361-4146-8d89-b7c870c70a36" />


## Proposal

Added the option for users to upload and preview PDF files within the editor.

Description:

- When a user types '/' on their keyboard, they can find an option called "PDF," which allows them to upload or embed PDF files.

virgile-dev commented:
Hey @dakshesh14 ! 
great to see you were able to give it a second shot. @AntoLC is on vacation right now. 
Once he is back he’ll able to review your PR. 
sorry for the delay  and thanks for your contribution

AntoLC commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
Seems to have some drag and drop issue with the embed tag, not sure why.

[- Docs (1).webm](https://github.com/user-attachments/assets/0c054063-b2a9-4ac1-b495-04a13e53219e)

dakshesh14 replied:
I am not sure either.

AntoLC commented on src/frontend/apps/impress/src/features/docs/doc-export/blocks-mapping/pdfPdf.tsx:
```diff
@@ -0,0 +1,31 @@
+    const pdfUrl = block.props.url;
+
+    return (
+      <View
+        style={{
+          marginVertical: 10,
+          padding: 10,
+          border: '1px solid #ccc',
+          borderRadius: 4,
+          backgroundColor: '#f9f9f9',
+        }}
+      >
+        <Text style={{ fontSize: 12, fontWeight: 'bold', marginBottom: 5 }}>
+          📄 {pdfName}
+        </Text>
+        {pdfUrl && (
+          <Text style={{ fontSize: 10, color: '#666' }}>Source: {pdfUrl}</Text>
+        )}
+        <Text style={{ fontSize: 10, color: '#999', marginTop: 5 }}>
+          [PDF content cannot be embedded in exported PDF]
+        </Text>
+      </View>
+    );
```
If it is not supported we should just fallback on what is made already, same for "Docx":
https://github.com/TypeCellOS/BlockNote/blob/7a66f11fd8a21ce3c1ffcb87590ae44b7f877ed6/packages/xl-pdf-exporter/src/pdf/defaultSchema/blocks.tsx#L205-L218

<img width="963" height="524" alt="image" src="https://github.com/user-attachments/assets/fa22ae6b-7d62-42c1-b4b6-0bb38a782e93" />

It is on the roadmap to improve the design export.

dakshesh14 replied:
I’m not sure I fully understand what you mean here. Could you clarify a bit more or give an example of how you’d like me to handle the fallback?

AntoLC replied:
You created this part for the PDF embed:
<img width="343" height="185" alt="image" src="https://github.com/user-attachments/assets/f1a1bee5-e8a0-440d-bd12-b5f32a570c6c" />

PDF embed are not supported, so let's keep using what we do already:
<img width="340" height="122" alt="image" src="https://github.com/user-attachments/assets/4efbb4d6-616d-43ea-9d2c-53cc06e926d3" />

dakshesh14 replied:
Done! Using the default for mapping PDF block.

AntoLC commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+        [editor, block],
+      );
+
+      const { wrapperRef, pdfWidth, handlePointerDown } = usePdfResizer(
+        block.props.previewWidth ?? 100,
+        handleResizeEnd,
+      );
+
+      return (
+        <div ref={contentRef} className="bn-file-block-content-wrapper">
+          {pdfUrl === '' ? (
+            <AddFileButton
+              block={block}
+              editor={editor as FileBlockEditor}
+              buttonText="Add PDF"
+              buttonIcon={<Icon iconName="upload" $size="18px" />}
+            />
+          ) : (
+            <FileBlockWrapper block={block} editor={editor as FileBlockEditor}>
```
Should be in another PR, but make `ResizableFileBlockWrapper` sizable on height as well will be great !

dakshesh14 replied:
Decided to revert changes related to resizing to make the PR short. We can look into resizing in a separate PR along with the height resizing option.

AntoLC commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+    },
+    aliases: ['pdf', 'document', 'embed', 'file'],
+    group,
+    icon: <Icon iconName="upload" $size="18px" />,
```
What about `picture_as_pdf` icon instead ?

AntoLC commented on docker/files/etc/nginx/conf.d/default.conf:
```diff
@@ -21,6 +21,9 @@ server {
         proxy_pass http://minio:9000/impress-media-storage/;
         proxy_set_header Host minio:9000;
 
+        proxy_hide_header Content-Disposition;
+        add_header Content-Disposition "inline";
+
```
I am not sure to understand why it is necessary, seems to work fine with the `embed` tag without changing to "inline" ?
Is there a browser where you get a problem maybe ?

dakshesh14 replied:
You are right, this was not required from my end as well.

dakshesh14 commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+            />
+          ) : (
+            <FileBlockWrapper block={block} editor={editor as FileBlockEditor}>
+              {isMobile ? (
+                <Box
+                  $display="flex"
+                  $align="center"
+                  $justify="center"
+                  $direction="row"
+                  $gap="1rem"
+                >
+                  <Icon iconName="picture_as_pdf" $size="18px" />
+                  <Box
+                    as="p"
+                    $display="inline-block"
+                    style={{
+                      textOverflow: 'ellipsis',
+                      whiteSpace: 'nowrap',
+                      overflow: 'hidden',
+                      display: 'inline',
+                    }}
+                  >
+                    {block.props.name}
+                  </Box>
+                </Box>
```
Yes, I added this because on iOS Safari the `<embed>` tag is supported, but it doesn’t render PDFs inline - it forces them to open in a new tab. Using `<object>` or `noembed` doesn’t help here, since the behavior is the same. Given that `<embed>` is still the most widely supported option across browsers (including iOS), I chose to stick with it.

But please let me know if you’d like me to take another look at this.

dakshesh14 commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+      showPreview: { default: true },
+      previewWidth: { default: undefined, type: 'number' },
+    },
+    isFileBlock: true,
```
Added `fileBlockAccept` to target only PDF files as suggested. Thanks for pointing it out.

dakshesh14 commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+            <AddFileButton
+              block={block}
+              editor={editor as FileBlockEditor}
+              buttonText="Add PDF"
```
Added translation here.

dakshesh14 commented on src/frontend/apps/impress/src/features/docs/doc-editor/components/custom-blocks/PdfBlock.tsx:
```diff
@@ -0,0 +1,151 @@
+                <Box
+                  $display="flex"
+                  $align="center"
+                  $justify="center"
```
Done! Using classes from blocknote for consistency with other embedded files.

dakshesh14 commented:
Hey @AntoLC,

I’ve addressed all your review comments and also resolved the merge conflict. I decided to handle PDF resizing in a separate PR to keep this one focused and shorter.

Could you please take another look when you get a chance?

AntoLC commented:
@dakshesh14 I took the freedom to finalize the PR to fit our needs (test e2e, manage loading upload with antivirus..).

Thank you very much for this contribution, it is a very nice feature  🎉 !! 
If you feel motivated to try to implement a height resizing, feel free to propose a contribution.

Thanks 🙏
